### PR TITLE
architecture as code: serverless blog post

### PR DESCRIPTION
**Pull request by spara**

### CODE REVIEW

File: content/blog/architecture-as-code-serverless/index.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  	egress=[{
+		'protocol': '-1',
+		'from_port': 0,
+		'to_port': 0,
+		'cidr_blocks': ['0.0.0.0/0'],
+	}]
+)
+
+# Create a load balancer to listen for HTTP traffic on port 80.
+alb = aws.lb.LoadBalancer('app-lb',
```
Comment (mikhailshilkov) [inline content/blog/architecture-as-code-serverless/index.md:61]: It's not quite clear to me why we need a load balancer in this context

File: content/blog/architecture-as-code-serverless/index.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	role=role.name,
+	policy_arn='arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy'
+)
+```
+
+To learn more about creating [reusable components]({{< relref "/docs/intro/concepts/programming-model#components" >}}), read about Pulumi's programming model, which shows how to author components.
+
+## Polyglot applications
+
+This example shows how to deploy multiple functions in Google Cloud Functions. Note that we have one function written in Python and another in Go. The code that deploys the function is written in python and creates a Buckets to hold the function code and deploys both functions in the same script.
```
Comment (mikhailshilkov) [inline content/blog/architecture-as-code-serverless/index.md:107]:
```suggestion
This example shows how to deploy multiple functions in Google Cloud Functions. Note that we have one function written in Python and another in Go. The code that deploys the function is written in python and creates a Bucket to hold the function code and deploys both functions in the same script.
```

File: content/blog/architecture-as-code-serverless/index.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+Despite the many advantages of serverless, one of the challenges of serverless is latency, more commonly known as “cold starts.” There are several ways to lessen the impact of cold starts, including:
+
+- warming is the practice of invoking additional workers to provide a pool of workers
+- provisioned concurrency has a set of workers are always warm and dedicated to a specific function
+- dynamic concurrency is similar to provisioned concurrency but the workers can autoscale as needed
+- scheduled profile is warming scheduled during peak use
+- autoscaling based utilization adds workers by measuring utilization
+
+Deploying infrastructure with code lets you implement these [strategies]({{< relref "/blog/aws-lambda-provisioned-concurrency-no-cold-starts" >}}) based on your application requirements. A [worked example](https://github.com/pulumi/examples/tree/master/aws-ts-serverless-raw) is available on Github.
```
Comment (mikhailshilkov) [inline content/blog/architecture-as-code-serverless/index.md:166]: Is this example related to cold starts?

File: content/blog/architecture-as-code-serverless/index.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    - serverless
+---
+
+In this fourth installment of Architecture as Code series, we’ll take a look at serverless, an architectural pattern that has quickly gained popularity among cloud practitioners. There are two reasons why serverless usage has proliferated: a cost-saving pay as you go model and elasticity that goes from zero to as many as needed to complete the task without managing servers.
+
+<!--more-->
+
+## Overview
+
+Serverless is a pattern for building services without having to manage servers. You write a service and deploy it on infrastructure managed by a cloud service provider. Serverless apps can be either a simple function invoked by an event or an application in a container. They are typically ephemeral, where the application runs within a fixed time or until it completes the task. Managed implementations such as AWS Lambda and Fargate, Azure Functions, and Google Cloud Run and Cloud Functions charge for the service only when it’s invoked and create instances, or workers, based on demand.
```
[the remark this answers is not in the document]
Comment (spara) [reply content/blog/architecture-as-code-serverless/index.md:18]: I think that the broader definition of serverless is run code or containers which puts Fargate in the serverless camp.

File: content/blog/architecture-as-code-serverless/index.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+Developers can write the application in any language that the service provider supports. That means developers can choose the language that best meets the needs of their application. For example, a compiled application may respond more quickly than a scripting language. Serverless applications can be either synchronous or asynchronous, depending on the task.
+
+Another component of serverless architecture is an API Gateway which maps HTTP operations such as GET, PUT, POST, and DELETE to functions. A gateway simplifies implementing REST interfaces for functions and applications.
```
[the remark this answers is not in the document]
Comment (spara) [reply content/blog/architecture-as-code-serverless/index.md:22]: revised to:

```suggestion
A common feature of serverless architecture is a REST interface which maps HTTP operations such as GET, PUT, POST, and DELETE to functions. A REST interface simplifies implementing REST interfaces for functions and applications. AWS offers API Gateway for building REST interfaces, while Azure and Google Cloud have out of the box solutions.
```

File: content/blog/architecture-as-code-serverless/index.md (hunk 1 above)
Comment (spara) [reply content/blog/architecture-as-code-serverless/index.md:61]: Pulling this from the example, more to show what is possible than practical.

File: content/blog/architecture-as-code-serverless/index.md (hunk 3 above)
Comment (spara) [reply content/blog/architecture-as-code-serverless/index.md:166]: added code sample for provisioned concurrency

File: content/blog/architecture-as-code-serverless/index.md [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+export("go_endpoint", go_function.https_trigger_url)
+```
+
+The [full example](https://github.com/pulumi/examples/tree/master/gcp-ts-serverless-raw) is available on Github.
+
+## Cold Starts
+
+Despite the many advantages of serverless, one of the challenges of serverless is latency, more commonly known as “cold starts.” There are several ways to lessen the impact of cold starts, including:
+
+- warming is the practice of invoking additional workers to provide a pool of workers
```
Comment (isaac-pulumi) [inline content/blog/architecture-as-code-serverless/index.md:160]:
```suggestion
- Invoking additional workers to provide a pool of workers - also known as 'warming'.
```

File: content/blog/architecture-as-code-serverless/index.md [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+- provisioned concurrency has a set of workers are always warm and dedicated to a specific function
```
Comment (isaac-pulumi) [inline content/blog/architecture-as-code-serverless/index.md:161]:
```suggestion
- Provisioning concurrent workers that are always warm and dedicated to a specific function.
```

File: content/blog/architecture-as-code-serverless/index.md [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+- dynamic concurrency is similar to provisioned concurrency but the workers can autoscale as needed
```
Comment (isaac-pulumi) [inline content/blog/architecture-as-code-serverless/index.md:162]:
```suggestion
- Dynamic concurrency - workers configured to autoscale as needed.
```

File: content/blog/architecture-as-code-serverless/index.md [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+- scheduled profile is warming scheduled during peak use
```
Comment (isaac-pulumi) [inline content/blog/architecture-as-code-serverless/index.md:163]:
```suggestion
- Scheduled profile - warming is applied during peak use.
```

File: content/blog/architecture-as-code-serverless/index.md [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+- autoscaling based utilization adds workers by measuring utilization
```
Comment (isaac-pulumi) [inline content/blog/architecture-as-code-serverless/index.md:164]:
```suggestion
- Autoscaling-based utilization - adds workers as service utilization increases.
```

File: content/blog/architecture-as-code-serverless/index.md [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+        functionName: lambda.name,
+        qualifier: lambda.version,
+        provisionedConcurrentExecutions,
+    });
+}
+```
+
+## Conclusion
+
+Serverless is an attractive alternative to architectures that require hands-on management of servers. Despite this, it does not relieve you of the tasks needed to create and manage the ancillary resources needed for a modern application in production. Architecture as code enables you to declare resources and reuse them to ensure a standard and reproducible infrastructure. Also, you can deploy multiple functions and applications with a single Pulumi project, saving time and effort required to deploy them individually. Finally, you can tune your deployment strategies to overcome cold-start issues. [Pulumi tutorials](https://www.pulumi.com/docs/tutorials/) can get you started with serverless deployments.
```
Comment (isaac-pulumi) [inline content/blog/architecture-as-code-serverless/index.md:201]:
```suggestion
Serverless is an attractive alternative to architectures that require hands-on management of servers. Despite this, it does not relieve you of the tasks needed to create and manage the ancillary resources needed for a modern application in production. Architecture as code enables you to declare resources and reuse them to ensure a standard and reproducible infrastructure. Also, you can deploy multiple functions and applications with a single Pulumi project, saving time and effort required to deploy them individually. Finally, you can tune your deployment strategies to overcome cold-start issues. [Pulumi tutorials](https://www.pulumi.com/docs/tutorials/) includes resources to help you get started with serverless deployments.
```

Comment (isaac-pulumi) [inline content/blog/architecture-as-code-serverless/index.md:201]: rel-ref needed
